File: Part 8/NTSA_timefreq_waveParamsSim.m
%%
%     COURSE: Solved problems in neural time series analysis
%    SECTION: Time-frequency analyses
%      VIDEO: Exploring wavelet parameters in simulated data
% Instructor: sincxpress.com
%
%%

%% simulate data as transient oscillation

pnts  = 4000;
srate = 1000;
stime  = (0:pnts-1)/srate - 2;

% gaussian parameters
fwhm = .4;

% sine wave parameters
sinefreq = 10; % for sine wave

% create signal
gaus   = exp( -(4*log(2)*stime.^2) / fwhm^2 );
cosw   = cos(2*pi*sinefreq*stime + 2*pi*rand);
signal = cosw .* gaus;

% get signal amplitude
sigamp = abs(hilbert(signal));


figure(1), clf, hold on
plot(stime,signal,'k','linew',3)
plot(stime,sigamp,'r--','linew',1)
xlabel('Time (s)')

%% comparing fixed number of wavelet cycles

% wavelet parameters
num_frex = 50;
min_freq =  2;
max_freq = 20;


% set a few different wavelet widths (FWHM parameter)
fwhms = [ .1 .5 2 ];

% other wavelet parameters
frex = linspace(min_freq,max_freq,num_frex);
wtime = -2:1/srate:2;
half_wave = (length(wtime)-1)/2;

% FFT parameters
nKern = length(wtime);
nConv = nKern+pnts-1;

% initialize output time-frequency data
tf = zeros(length(fwhms),length(frex),pnts);


% FFT of data (doesn't change on frequency iteration)
dataX = fft( signal ,nConv);


% loop over cycles
for fwhmi=1:length(fwhms)
    
    for fi=1:length(frex)
        
        % create wavelet and get its FFT
        cmw  = exp(2*1i*pi*frex(fi).*wtime) .* exp(-4*log(2)*wtime.^2./fwhms(fwhmi)^2);
        cmwX = fft(cmw,nConv);
        cmwX = cmwX./max(cmwX);
        
        % run convolution, trim edges, and reshape to 2D (time X trials)
        as = ifft(cmwX.*dataX);
        as = as(half_wave+1:end-half_wave);
        
        % put power data into big matrix
        tf(fwhmi,fi,:) = abs(as);
    end
end


%% plot results

figure(3), clf, colormap parula

% time-frequency plots
for fwhmi=1:length(fwhms)
    subplot(2,3,fwhmi)
    contourf(stime,frex,squeeze(tf(fwhmi,:,:)),40,'linecolor','none')
    title([ 'Wavelet with ' num2str(fwhms(fwhmi)) ' s FWHM' ])
    xlabel('Time (s)'), ylabel('Frequency (Hz)')
end


% show amplitude envelopes at peak frequency
fidx = dsearchn(frex',sinefreq);
subplot(223), hold on
plot(stime,squeeze(tf(:,fidx,:)))
plot(stime,sigamp/2,'k--','linew',2)
xlabel('Time (s)')
legend([regexp(num2str(fwhms),' +','split'),'Truth'])
title('Time domain')


% show spectra at one time point
tidx = dsearchn(stime',0);
subplot(224), hold on
plot(frex,squeeze(tf(:,:,tidx)))
plot(linspace(0,srate,pnts),2*abs(fft(signal)/pnts),'k--','linew',2)
set(gca,'xlim',frex([1 end]))
xlabel('Frequency (Hz)')
legend([regexp(num2str(fwhms),' +','split'),'Truth'])
title('Frequency domain')

%% done.

File: Part 8/NTSA_timefreq_waveParamsSim.py
# %%
#     COURSE: Solved problems in neural time series analysis
#    SECTION: Time-frequency analyses
#    TEACHER: Mike X Cohen, sincxpress.com
#      VIDEO: Exploring wavelet parameters in simulated data

# !%matplotlib inline
import os
import sys

import matplotlib.pyplot as plt
import numpy as np
from numpy.fft import fft, ifft
from scipy.signal import hilbert

# Set figure settings
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from utils import get_cmap, set_fig_dpi, set_style

set_fig_dpi(), set_style()
cmap = get_cmap('parula')

# %% Simulate data as transient oscillation

pnts = 4000
srate = 1000
stime = np.arange(0, pnts) / srate - 2

# Gaussian parameters
fwhm = 0.4

# Sine wave parameters
sine_freq = 10

# Create signal
gaus = np.exp(-(4 * np.log(2) * stime**2) / fwhm**2)
cosw = np.cos(2 * np.pi * sine_freq * stime + 2 * np.pi * np.random.rand())
signal = cosw * gaus

# Get signal amplitude
sig_amp = np.abs(hilbert(signal))

plt.figure()
plt.plot(stime, signal, 'k')
plt.plot(stime, sig_amp, 'r--', linewidth=1)
plt.xlabel('Time (s)')
plt.xlabel('Amplitude (a.u.)')

plt.show()

# %% Comparing fixed number of wavelet cycles

# Wavelet parameters
num_freqs = 50
min_freq = 2
max_freq = 20

# Set a few different wavelet widths (FWHM parameter)
fwhms = [0.1, 0.5, 2]

# Other wavelet parameters
freqs = np.linspace(min_freq, max_freq, num_freqs)
w_time = np.arange(-2, 2 + 1 / srate, 1 / srate)
half_wave = (len(w_time) - 1) // 2

# FFT parameters
n_kern = len(w_time)
n_conv = n_kern + pnts - 1

# Initialize output time-frequency data
tf = np.zeros((len(fwhms), len(freqs), pnts))

# FFT of data (doesn't change on frequency iteration)
dataX = fft(signal, n_conv)

# Loop over cycles
for fwhmi in range(len(fwhms)):
    for fi in range(len(freqs)):
        # Create wavelet and get its FFT
        cmw = np.exp(2 * 1j * np.pi * freqs[fi] * w_time) * np.exp(
            -4 * np.log(2) * w_time**2 / fwhms[fwhmi] ** 2
        )
        cmwX = fft(cmw, n_conv)
        cmwX = cmwX / cmwX[np.argmax(np.abs(cmwX))]

        # Run convolution, trim edges
        comp_sig = ifft(cmwX * dataX)
        comp_sig = comp_sig[half_wave:-half_wave]

        # Put power data into big matrix
        tf[fwhmi, fi, :] = np.abs(comp_sig)

# %% Plot results

plt.figure()

# Define the grid
gridsize = (4, 6)  # Rows, columns
ax1 = plt.subplot2grid(gridsize, (0, 0), colspan=2, rowspan=2)
ax2 = plt.subplot2grid(gridsize, (0, 2), colspan=2, rowspan=2)
ax3 = plt.subplot2grid(gridsize, (0, 4), colspan=2, rowspan=2)
ax4 = plt.subplot2grid(gridsize, (2, 0), colspan=3, rowspan=2)
ax5 = plt.subplot2grid(gridsize, (2, 3), colspan=3, rowspan=2)

# Time-frequency plots
ax1.contourf(stime, freqs, tf[0, :, :], 40, cmap=cmap)
ax1.set_title(f'{fwhms[0]} s FWHM')
ax1.set_xlabel('Time (s)')
ax1.set_ylabel('Frequency (Hz)')

ax2.contourf(stime, freqs, tf[1, :, :], 40, cmap=cmap)
ax2.set_title(f'{fwhms[1]} s FWHM')
ax2.set_xlabel('Time (s)')

ax3.contourf(stime, freqs, tf[2, :, :], 40, cmap=cmap)
ax3.set_title(f'{fwhms[2]} s FWHM')
ax3.set_xlabel('Time (s)')


# Show amplitude envelopes at peak frequency
f_idx = np.argmin(np.abs(freqs - sine_freq))
ax4.plot(stime, tf[:, f_idx, :].T)
ax4.plot(stime, sig_amp / 2, 'k--')
ax4.set_title('Time domain')
ax4.set_xlabel('Time (s)')
ax4.set_ylabel('Amplitude (a.u.)')
ax4.legend([str(fwhm) for fwhm in fwhms] + ['Truth'])

# Show spectra at one time point
t_idx = np.argmin(np.abs(stime))
ax5.plot(freqs, tf[:, :, t_idx].T)
ax5.plot(np.linspace(0, srate, pnts), 2 * np.abs(fft(signal) / pnts), 'k--')
ax5.set_xlim([freqs[0], freqs[-1]])
ax5.set_xlabel('Frequency (Hz)')
ax5.set_title('Frequency domain')
ax5.legend([str(fwhm) for fwhm in fwhms] + ['Truth'])

plt.tight_layout()
plt.show()

# %%
